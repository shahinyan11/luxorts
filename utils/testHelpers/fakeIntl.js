import { is } from 'ramda';

const formatMessage = ({ id, values }) => {
  let valuesPart = '';
  if (values) {
    valuesPart = ', values: {';
    valuesPart = Object.entries(values).reduce(
      (result, entry) =>
        `${result} ${entry[0]}: ${is(String)(entry[1]) ? `'${entry[1]}'` : entry[1]},`,
      valuesPart,
    );
    valuesPart = `${valuesPart.slice(0, -1)} }`;
  }
  return `{Translation id: ${id}${valuesPart}}`;
};

const fakeIntl = {
  formatMessage,
};

export default fakeIntl;
