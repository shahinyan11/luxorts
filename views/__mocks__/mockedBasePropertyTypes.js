export default [
  {
    value: '51e35121-c27e-46fc-9feb-b51dab2490ee',
    label: 'repudiandae',
    description: 'description repudiandae',
  },
  {
    value: '995c5714-185d-4085-8512-690903334f6e',
    label: 'molestias',
    description: 'description molestias',
  },
];
