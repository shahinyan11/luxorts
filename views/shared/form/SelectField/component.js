import PropTypes from 'prop-types';
import { Form, Select } from 'antd';
import { FormattedMessage } from 'react-intl';
import { last } from 'ramda';
import { isArray } from 'lodash';

import useFormattedOrRawMessage from 'hooks/useFormattedOrRawMessage';

import InputHelp from 'views/shared/form/InputHelp';
import SvgIcon from 'views/shared/SvgIcon';

import SelectOption from './SelectOption';
import useContainer from './hook';

const SelectField = ({
  label,
  name,
  formItemClassName,
  placeholder,
  tooltip,
  options,
  optionComponent: Option,
  ...props
}) => {
  const formattedOrRawMessage = useFormattedOrRawMessage();
  const { field, meta, onChangeHandler, onBlurHandler } = useContainer({ name });
  const hasError = meta.touched && meta.error;
  const Error = hasError && (
    <InputHelp type="error" text={isArray(meta.error) ? last(meta.error) : meta.error} />
  );
  const Tooltip = tooltip && <InputHelp text={tooltip} />;

  return (
    <Form.Item
      className={formItemClassName}
      label={label && <FormattedMessage {...label} />}
      htmlFor={name}
      validateStatus={hasError && 'error'}
      help={Error || Tooltip}
    >
      <Select
        {...props}
        id={name}
        value={field.value}
        onChange={onChangeHandler}
        onBlur={onBlurHandler}
        placeholder={formattedOrRawMessage(placeholder)}
        suffixIcon={<SvgIcon icon="arrow-down" />}
      >
        {options.map(Option)}
      </Select>
    </Form.Item>
  );
};

SelectField.defaultProps = {
  formItemClassName: undefined,
  label: undefined,
  placeholder: undefined,
  tooltip: undefined,
  options: [],
  optionComponent: SelectOption,
};

SelectField.propTypes = {
  name: PropTypes.string.isRequired,
  formItemClassName: PropTypes.string,
  label: PropTypes.shape({
    id: PropTypes.string.isRequired,
    values: PropTypes.shape(),
  }),
  placeholder: PropTypes.oneOfType([PropTypes.shape(), PropTypes.string]),
  tooltip: PropTypes.oneOfType([PropTypes.shape(), PropTypes.string]),
  options: PropTypes.arrayOf(PropTypes.shape()),
  optionComponent: PropTypes.elementType,
};

export default SelectField;
