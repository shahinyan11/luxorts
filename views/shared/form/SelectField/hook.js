import { useField } from 'formik';

import useImmutableCallback from 'hooks/useImmutableCallback';

function useContainer({ name }) {
  const [field, meta, helpers] = useField(name);
  const { setValue, setTouched } = helpers;

  /**
   * Handle blur
   */
  const onBlurHandler = useImmutableCallback(() => {
    setTouched(true);
  });

  /**
   * Handle select
   */
  const onChangeHandler = useImmutableCallback((value) => {
    setValue(value);
  });

  return {
    field,
    meta,
    onChangeHandler,
    onBlurHandler,
  };
}

export default useContainer;
