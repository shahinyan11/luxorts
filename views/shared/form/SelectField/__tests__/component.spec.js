import { shallow } from 'enzyme';
import { mergeDeepRight } from 'ramda';

import fakeIntl from 'utils/testHelpers/fakeIntl';
import mockedBasePropertyTypes from 'views/__mocks__/mockedBasePropertyTypes';

import useContainer from '../hook';
import SelectField from '../component';

const mockedHookData = {
  intl: fakeIntl,
  meta: {
    error: null,
    touched: false,
  },
  field: { value: 'value' },
  onChangeHandler: jest.fn(),
  onBlurHandler: jest.fn(),
};
jest.mock('../hook', () => jest.fn(() => mockedHookData));

describe('SelectField component', () => {
  const defaultProps = {
    name: 'name',
    options: mockedBasePropertyTypes,
  };

  let component = null;

  beforeEach(() => {
    component = shallow(<SelectField {...defaultProps} />);
  });

  it('matches snapshot', () => {
    expect(component).toMatchSnapshot();
  });

  it('matches snapshot when optionComponent is present', () => {
    component.setProps({
      optionComponent: ({ value }) => <div key={value} />,
    });

    expect(component).toMatchSnapshot();
  });

  it('matches snapshot when options aren`t present', () => {
    component.setProps({
      options: undefined,
    });

    expect(component).toMatchSnapshot();
  });

  it('matches snapshot when formItemClassName is present', () => {
    component.setProps({
      formItemClassName: 'test',
    });

    expect(component).toMatchSnapshot();
  });

  it('matches snapshot when label is present', () => {
    component.setProps({
      label: { id: 'label.id' },
    });

    expect(component).toMatchSnapshot();
  });

  it('matches snapshot when placeholder is present', () => {
    component.setProps({
      placeholder: { id: 'placeholder.id' },
    });

    expect(component).toMatchSnapshot();
  });

  it('matches snapshot when tooltip is present', () => {
    component.setProps({
      tooltip: { id: 'tooltip.id' },
    });

    expect(component).toMatchSnapshot();
  });

  it('matches snapshot when touched and has no error', () => {
    useContainer.mockReturnValueOnce(
      mergeDeepRight(mockedHookData, {
        meta: {
          touched: true,
        },
      }),
    );

    component = shallow(<SelectField {...defaultProps} />);

    expect(component).toMatchSnapshot();
  });

  it('matches snapshot when touched and has error', () => {
    useContainer.mockReturnValueOnce(
      mergeDeepRight(mockedHookData, {
        meta: {
          touched: true,
          error: { id: 'error.id' },
        },
      }),
    );

    component = shallow(<SelectField {...defaultProps} />);

    expect(component).toMatchSnapshot();
  });

  it('matches snapshot when touched and has error and tooltip is present', () => {
    useContainer.mockReturnValueOnce(
      mergeDeepRight(mockedHookData, {
        meta: {
          touched: true,
          error: { id: 'error.id' },
        },
      }),
    );

    const props = {
      ...defaultProps,
      tooltip: { id: 'tooltip.id' },
    };

    component = shallow(<SelectField {...props} />);

    expect(component).toMatchSnapshot();
  });
});
