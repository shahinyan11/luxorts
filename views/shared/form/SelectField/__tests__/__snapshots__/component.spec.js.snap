// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectField component matches snapshot 1`] = `
<FormItem
  htmlFor="name"
  validateStatus={false}
>
  <ForwardRef(InternalSelect)
    id="name"
    onBlur={[MockFunction]}
    onChange={[MockFunction]}
    suffixIcon={
      <SvgIcon
        icon="arrow-down"
      />
    }
    value="value"
  >
    <Option
      key="51e35121-c27e-46fc-9feb-b51dab2490ee"
      label={
        <FormattedOrRawMessage
          message="repudiandae"
        />
      }
      value="51e35121-c27e-46fc-9feb-b51dab2490ee"
    >
      <FormattedOrRawMessage
        message="repudiandae"
      />
    </Option>
    <Option
      key="995c5714-185d-4085-8512-690903334f6e"
      label={
        <FormattedOrRawMessage
          message="molestias"
        />
      }
      value="995c5714-185d-4085-8512-690903334f6e"
    >
      <FormattedOrRawMessage
        message="molestias"
      />
    </Option>
  </ForwardRef(InternalSelect)>
</FormItem>
`;

exports[`SelectField component matches snapshot when formItemClassName is present 1`] = `
<FormItem
  className="test"
  htmlFor="name"
  validateStatus={false}
>
  <ForwardRef(InternalSelect)
    id="name"
    onBlur={[MockFunction]}
    onChange={[MockFunction]}
    suffixIcon={
      <SvgIcon
        icon="arrow-down"
      />
    }
    value="value"
  >
    <Option
      key="51e35121-c27e-46fc-9feb-b51dab2490ee"
      label={
        <FormattedOrRawMessage
          message="repudiandae"
        />
      }
      value="51e35121-c27e-46fc-9feb-b51dab2490ee"
    >
      <FormattedOrRawMessage
        message="repudiandae"
      />
    </Option>
    <Option
      key="995c5714-185d-4085-8512-690903334f6e"
      label={
        <FormattedOrRawMessage
          message="molestias"
        />
      }
      value="995c5714-185d-4085-8512-690903334f6e"
    >
      <FormattedOrRawMessage
        message="molestias"
      />
    </Option>
  </ForwardRef(InternalSelect)>
</FormItem>
`;

exports[`SelectField component matches snapshot when label is present 1`] = `
<FormItem
  htmlFor="name"
  label={
    <Memo(FormattedMessage)
      id="label.id"
    />
  }
  validateStatus={false}
>
  <ForwardRef(InternalSelect)
    id="name"
    onBlur={[MockFunction]}
    onChange={[MockFunction]}
    suffixIcon={
      <SvgIcon
        icon="arrow-down"
      />
    }
    value="value"
  >
    <Option
      key="51e35121-c27e-46fc-9feb-b51dab2490ee"
      label={
        <FormattedOrRawMessage
          message="repudiandae"
        />
      }
      value="51e35121-c27e-46fc-9feb-b51dab2490ee"
    >
      <FormattedOrRawMessage
        message="repudiandae"
      />
    </Option>
    <Option
      key="995c5714-185d-4085-8512-690903334f6e"
      label={
        <FormattedOrRawMessage
          message="molestias"
        />
      }
      value="995c5714-185d-4085-8512-690903334f6e"
    >
      <FormattedOrRawMessage
        message="molestias"
      />
    </Option>
  </ForwardRef(InternalSelect)>
</FormItem>
`;

exports[`SelectField component matches snapshot when optionComponent is present 1`] = `
<FormItem
  htmlFor="name"
  validateStatus={false}
>
  <ForwardRef(InternalSelect)
    id="name"
    onBlur={[MockFunction]}
    onChange={[MockFunction]}
    suffixIcon={
      <SvgIcon
        icon="arrow-down"
      />
    }
    value="value"
  >
    <div
      key="51e35121-c27e-46fc-9feb-b51dab2490ee"
    />
    <div
      key="995c5714-185d-4085-8512-690903334f6e"
    />
  </ForwardRef(InternalSelect)>
</FormItem>
`;

exports[`SelectField component matches snapshot when options aren\`t present 1`] = `
<FormItem
  htmlFor="name"
  validateStatus={false}
>
  <ForwardRef(InternalSelect)
    id="name"
    onBlur={[MockFunction]}
    onChange={[MockFunction]}
    suffixIcon={
      <SvgIcon
        icon="arrow-down"
      />
    }
    value="value"
  />
</FormItem>
`;

exports[`SelectField component matches snapshot when placeholder is present 1`] = `
<FormItem
  htmlFor="name"
  validateStatus={false}
>
  <ForwardRef(InternalSelect)
    id="name"
    onBlur={[MockFunction]}
    onChange={[MockFunction]}
    placeholder="{Translation id: placeholder.id}"
    suffixIcon={
      <SvgIcon
        icon="arrow-down"
      />
    }
    value="value"
  >
    <Option
      key="51e35121-c27e-46fc-9feb-b51dab2490ee"
      label={
        <FormattedOrRawMessage
          message="repudiandae"
        />
      }
      value="51e35121-c27e-46fc-9feb-b51dab2490ee"
    >
      <FormattedOrRawMessage
        message="repudiandae"
      />
    </Option>
    <Option
      key="995c5714-185d-4085-8512-690903334f6e"
      label={
        <FormattedOrRawMessage
          message="molestias"
        />
      }
      value="995c5714-185d-4085-8512-690903334f6e"
    >
      <FormattedOrRawMessage
        message="molestias"
      />
    </Option>
  </ForwardRef(InternalSelect)>
</FormItem>
`;

exports[`SelectField component matches snapshot when tooltip is present 1`] = `
<FormItem
  help={
    <InputHelp
      icon="alert"
      text={
        Object {
          "id": "tooltip.id",
        }
      }
    />
  }
  htmlFor="name"
  validateStatus={false}
>
  <ForwardRef(InternalSelect)
    id="name"
    onBlur={[MockFunction]}
    onChange={[MockFunction]}
    suffixIcon={
      <SvgIcon
        icon="arrow-down"
      />
    }
    value="value"
  >
    <Option
      key="51e35121-c27e-46fc-9feb-b51dab2490ee"
      label={
        <FormattedOrRawMessage
          message="repudiandae"
        />
      }
      value="51e35121-c27e-46fc-9feb-b51dab2490ee"
    >
      <FormattedOrRawMessage
        message="repudiandae"
      />
    </Option>
    <Option
      key="995c5714-185d-4085-8512-690903334f6e"
      label={
        <FormattedOrRawMessage
          message="molestias"
        />
      }
      value="995c5714-185d-4085-8512-690903334f6e"
    >
      <FormattedOrRawMessage
        message="molestias"
      />
    </Option>
  </ForwardRef(InternalSelect)>
</FormItem>
`;

exports[`SelectField component matches snapshot when touched and has error 1`] = `
<FormItem
  help={
    <InputHelp
      icon="alert"
      text={
        Object {
          "id": "error.id",
        }
      }
      type="error"
    />
  }
  htmlFor="name"
  validateStatus="error"
>
  <ForwardRef(InternalSelect)
    id="name"
    onBlur={[MockFunction]}
    onChange={[MockFunction]}
    suffixIcon={
      <SvgIcon
        icon="arrow-down"
      />
    }
    value="value"
  >
    <Option
      key="51e35121-c27e-46fc-9feb-b51dab2490ee"
      label={
        <FormattedOrRawMessage
          message="repudiandae"
        />
      }
      value="51e35121-c27e-46fc-9feb-b51dab2490ee"
    >
      <FormattedOrRawMessage
        message="repudiandae"
      />
    </Option>
    <Option
      key="995c5714-185d-4085-8512-690903334f6e"
      label={
        <FormattedOrRawMessage
          message="molestias"
        />
      }
      value="995c5714-185d-4085-8512-690903334f6e"
    >
      <FormattedOrRawMessage
        message="molestias"
      />
    </Option>
  </ForwardRef(InternalSelect)>
</FormItem>
`;

exports[`SelectField component matches snapshot when touched and has error and tooltip is present 1`] = `
<FormItem
  help={
    <InputHelp
      icon="alert"
      text={
        Object {
          "id": "error.id",
        }
      }
      type="error"
    />
  }
  htmlFor="name"
  validateStatus="error"
>
  <ForwardRef(InternalSelect)
    id="name"
    onBlur={[MockFunction]}
    onChange={[MockFunction]}
    suffixIcon={
      <SvgIcon
        icon="arrow-down"
      />
    }
    value="value"
  >
    <Option
      key="51e35121-c27e-46fc-9feb-b51dab2490ee"
      label={
        <FormattedOrRawMessage
          message="repudiandae"
        />
      }
      value="51e35121-c27e-46fc-9feb-b51dab2490ee"
    >
      <FormattedOrRawMessage
        message="repudiandae"
      />
    </Option>
    <Option
      key="995c5714-185d-4085-8512-690903334f6e"
      label={
        <FormattedOrRawMessage
          message="molestias"
        />
      }
      value="995c5714-185d-4085-8512-690903334f6e"
    >
      <FormattedOrRawMessage
        message="molestias"
      />
    </Option>
  </ForwardRef(InternalSelect)>
</FormItem>
`;

exports[`SelectField component matches snapshot when touched and has no error 1`] = `
<FormItem
  htmlFor="name"
  validateStatus={null}
>
  <ForwardRef(InternalSelect)
    id="name"
    onBlur={[MockFunction]}
    onChange={[MockFunction]}
    suffixIcon={
      <SvgIcon
        icon="arrow-down"
      />
    }
    value="value"
  >
    <Option
      key="51e35121-c27e-46fc-9feb-b51dab2490ee"
      label={
        <FormattedOrRawMessage
          message="repudiandae"
        />
      }
      value="51e35121-c27e-46fc-9feb-b51dab2490ee"
    >
      <FormattedOrRawMessage
        message="repudiandae"
      />
    </Option>
    <Option
      key="995c5714-185d-4085-8512-690903334f6e"
      label={
        <FormattedOrRawMessage
          message="molestias"
        />
      }
      value="995c5714-185d-4085-8512-690903334f6e"
    >
      <FormattedOrRawMessage
        message="molestias"
      />
    </Option>
  </ForwardRef(InternalSelect)>
</FormItem>
`;
